library/ssl_client.c

```diff
@@ -843,6 +843,38 @@ static int ssl_prepare_client_hello( mbedtls_ssl_context *ssl )
         }
     }
 
+#if defined(MBEDTLS_SSL_PROTO_TLS1_3) && \
+    defined(MBEDTLS_SSL_SERVER_NAME_INDICATION)
+    if( ssl->tls_version == MBEDTLS_SSL_VERSION_TLS1_3  &&
+        ssl->handshake->resume )
+    {
+        int hostname_mismatch = 0;
+        if( ssl->session_negotiate->hostname != NULL )
+        {
+            if( ssl->hostname != NULL )
+            {
+                if( strcmp( ssl->hostname, ssl->session_negotiate->hostname) )
+                    hostname_mismatch = 1;
+            }
+            else
+                hostname_mismatch = 1;
+        }
+        else
+            hostname_mismatch = ssl->hostname != NULL;
```

Comment:
```suggestion
        int hostname_mismatch = ssl->hostname != NULL ||
                                ssl->session_negotiate->hostname != NULL;
        if( ssl->hostname != NULL && ssl->session_negotiate->hostname != NULL )
            hostname_mismatch = strcmp( ... ) != 0;
```
This is probably better, based on @yuhaoth proposal (see https://github.com/Mbed-TLS/mbedtls/pull/6299#discussion_r991296594).